### PR TITLE
core: add detail spec for evaluate result

### PR DESCRIPTION
**Pull request by rickycao-qy**
Refer to issue #585 

We need to support the evaluation result for both the train and test dataset. This PR is to support this case from the view of core spec. Meanwhile, we will change the plugins to support this feature.

### CODE REVIEW

**Comment by yorkie** [comment]
Shall we change the database as well?

File: packages/core/src/types/other.ts [hunk 1]
```diff
@@ -3,11 +3,17 @@ export interface Statistic {
...
 }
 
-export interface EvaluateResult {
+export interface EvaluateResultItem {
   pass?: boolean;
   [key: string]: any;
 }
 
+export interface EvaluateResult {
+  trainResult?: EvaluateResultItem;
```
Comment (yorkie) [inline packages/core/src/types/other.ts:12]:
```suggestion
  train?: EvaluateResultItem;
```

The result might be duplicated with the interface name?

**Comment by yorkie** [comment]
> Meanwhile, we will change the plugins to support this feature.

How does a plugin will write the train/test result?

**Comment by rickycao-qy** [comment]
> > Meanwhile, we will change the plugins to support this feature.
> 
> How does a plugin will write the train/test result?

Have changed the type definition to support compatibility for current plugins. Later on, the plugin can be re-written to support new type definition

File: packages/core/src/types/other.ts [hunk 2]
```diff
@@ -3,11 +3,19 @@ export interface Statistic {
   metricValue: number;
 }
 
-export interface EvaluateResult {
+export interface EvaluateResultItem {
   pass?: boolean;
   [key: string]: any;
 }
 
+export interface EvaluateResultDataset {
```
Comment (yorkie) [inline packages/core/src/types/other.ts:11]:
```suggestion
export interface EvaluateResultOnDatasets {
```

**Comment by yorkie** [comment]
> the plugin can be re-written to support new type definition

How does the plugin to write the backward-compatible for this feature? For example, a new plugin returns `EvaluateResultOnDatasets` on an older Pipcook, shall we have a way to set up the plugin could be installed and ran for the current daemon?

**Comment by yorkie** [comment]
And see https://coveralls.io/builds/34075298, the code coverage is decreased -1.2% at the pipeline.ts.

**Comment by rickycao-qy** [comment]
> > the plugin can be re-written to support new type definition
> 
> How does the plugin to write the backward-compatible for this feature? For example, a new plugin returns `EvaluateResultOnDatasets` on an older Pipcook, shall we have a way to set up the plugin could be installed and ran for the current daemon?

To stop installing plugins seems not a good experience. I'm thinking if it's a good idea to specify `EvaluateResult = EvaluateResultItem & EvaluateResultDataset` since the old version just has `[key: string]: any`. In this case,  if a new plugin runs on the old daemon, it still conforms to this definition.

**Comment by yorkie** [comment]
> To stop installing plugins seems not a good experience.

What I meant is just to verify if they can be installed on the host.

> I'm thinking if it's a good idea to specify EvaluateResult = EvaluateResultItem & EvaluateResultDataset since the old version just has [key: string]: any

Exactly, that's really the bad parts of using the dynamical definition before.

**Review by yorkie — APPROVED** [review]

**Review by WenheLI — APPROVED** [review]
LGTM

**Pull request merged**